File: cabot/cabot/handle_v2.py
#!/usr/bin/env python

# Copyright (c) 2020, 2022  Carnegie Mellon University
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from rclpy.duration import Duration
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from rclpy.qos import qos_profile_sensor_data
from cabot import button
from std_msgs.msg import Bool
from std_msgs.msg import String
from std_msgs.msg import UInt8
import time


class Handle:
    UNKNOWN = 0
    LEFT_TURN = 1
    RIGHT_TURN = 2
    LEFT_DEV = 3
    RIGHT_DEV = 4
    FRONT = 5
    LEFT_ABOUT_TURN = 6
    RIGHT_ABOUT_TURN = 7
    BUTTON_CLICK = 8
    BUTTON_HOLDDOWN = 9
    STIMULI_COUNT = 10
    stimuli_names = ["unknown", "left_turn", "right_turn", "left_dev", "right_dev",
                     "front", "left_about_turn", "right_about_turn", "button_click", "button_holddown"]

    @staticmethod
    def get_name(stimulus):
        return Handle.stimuli_names[stimulus]

    def __init__(self, node, realworld_use=True, event_listener=None, button_keys=[]):
        self.node = node
        self.logger = node.get_logger()
        self.event_listener = event_listener
        self.button_keys = button_keys
        self.number_of_buttons = len(button_keys)
        self.lastUp = [None]*self.number_of_buttons
        self.lastDwn = [None]*self.number_of_buttons
        self.upCount = [0]*self.number_of_buttons
        self.btnDwn = [False]*self.number_of_buttons
        self.power = 255
        self.vibrator1_pub = node.create_publisher(UInt8, '/cabot/vibrator1', 100)
        self.vibrator2_pub = node.create_publisher(UInt8, '/cabot/vibrator2', 100)
        self.vibrator3_pub = node.create_publisher(UInt8, '/cabot/vibrator3', 100)
        self.vibrator4_pub = node.create_publisher(UInt8, '/cabot/vibrator4', 100)
        for i in range(0, self.number_of_buttons):
            _ = node.create_subscription(Bool, F"/cabot/pushed_{i+1}",
                                         lambda msg, i=i: self.button_callback(msg, i), qos_profile_sensor_data, callback_group=MutuallyExclusiveCallbackGroup())

        self.duration = 15
        self.duration_single_vibration = 40
        self.duration_about_turn = 40
        self.duration_button_click = 5
        self.duration_button_holddown = 10
        self.sleep = 150
        self.updown = True
        self.num_vibrations_turn = 4
        self.num_vibrations_deviation = 2
        self.num_vibrations_about_turn = 2
        self.num_vibrations_confirmation = 1
        self.num_vibrations_button_click = 1
        self.num_vibrations_button_holddown = 1

        self.callbacks = [None]*Handle.STIMULI_COUNT
        self.callbacks[Handle.LEFT_TURN] = self.vibrate_left_turn
        self.callbacks[Handle.RIGHT_TURN] = self.vibrate_right_turn
        self.callbacks[Handle.LEFT_DEV] = self.vibrate_left_deviation
        self.callbacks[Handle.RIGHT_DEV] = self.vibrate_right_deviation
        self.callbacks[Handle.FRONT] = self.vibrate_front
        self.callbacks[Handle.LEFT_ABOUT_TURN] = self.vibrate_about_left_turn
        self.callbacks[Handle.RIGHT_ABOUT_TURN] = self.vibrate_about_right_turn
        self.callbacks[Handle.BUTTON_CLICK] = self.vibrate_button_click
        self.callbacks[Handle.BUTTON_HOLDDOWN] = self.vibrate_button_holddown

        self.eventSub = node.create_subscription(String, '/cabot/event', self.event_callback, 10, callback_group=MutuallyExclusiveCallbackGroup())

        self.double_click_interval = Duration(seconds=0.25)
        self.ignore_interval = Duration(seconds=0.05)
        self.holddown_interval = Duration(seconds=3.0)

    def button_callback(self, msg, index):
        self._button_check(msg, button.__dict__[self.button_keys[index]], index)

    def _button_check(self, msg, key, index):
        event = None
        now = self.node.get_clock().now()
        # detect change from button up to button down to emit a button down event
        if msg.data and not self.btnDwn[index] and \
           not (self.lastUp[index] is not None and
                now - self.lastUp[index] < self.ignore_interval):
            event = {"button": key, "up": False}
            self.btnDwn[index] = True
            self.lastDwn[index] = now # for holddown detection
        # detect change from button down to button up to emit a button up event
        if not msg.data and self.btnDwn[index]:
            event = {"button": key, "up": True}
            self.upCount[index] += 1
            self.lastUp[index] = now
            self.btnDwn[index] = False
        # detect the end of a series of button down/up events to emit a click event
        if self.lastUp[index] is not None and \
           not self.btnDwn[index] and \
           now - self.lastUp[index] > self.double_click_interval:
            # emit click event only if a hold down event is not detected
            if self.lastDwn[index] is not None:
                event = {"buttons": key, "count": self.upCount[index]}
            self.lastUp[index] = None
            self.upCount[index] = 0
        # detect button hold down to emit a holddown event
        if msg.data and self.btnDwn[index] and \
           (self.lastDwn[index] is not None and \
           now - self.lastDwn[index] > self.holddown_interval):
            event = {"holddown":key}
            # clear lastDwn[index] after holddown event emission to prevent multiple event emissions
            self.lastDwn[index] = None

        if event is not None:
            if self.event_listener is not None:
                self.event_listener(event)

    def event_callback(self, msg):
        name = msg.data
        if name in Handle.stimuli_names:
            index = Handle.stimuli_names.index(name)
            self.execute_stimulus(index)

    def execute_stimulus(self, index):
        self.logger.info("execute_stimulus, %d" % (index))
        callback = self.callbacks[index]
        if callback is not None:
            callback()
            self.logger.info("executed")

    def vibrate(self, pub):
        msg = UInt8()
        msg.data = self.power
        pub.publish(msg)

    def stop(self, pub):
        msg = UInt8()
        msg.data = 0
        pub.publish(msg)

    def vibrate_all(self, time):
        msg = UInt8()
        msg.data = self.power
        self.vibrator1_pub.publish(msg)
        self.vibrator2_pub.publish(msg)
        self.vibrator3_pub.publish(msg)
        self.vibrator4_pub.publish(msg)

        time.sleep(time)

        msg.data = 0
        self.vibrator1_pub.publish(msg)
        self.vibrator2_pub.publish(msg)
        self.vibrator3_pub.publish(msg)
        self.vibrator4_pub.publish(msg)

    def vibrate_left_turn(self):
        if self.updown:
            self.vibrate_pattern(self.vibrator3_pub, self.num_vibrations_turn, self.duration)
        else:
            self.vibrate_pattern(self.vibrator4_pub, self.num_vibrations_turn, self.duration)

    def vibrate_right_turn(self):
        if self.updown:
            self.vibrate_pattern(self.vibrator4_pub, self.num_vibrations_turn, self.duration)
        else:
            self.vibrate_pattern(self.vibrator3_pub, self.num_vibrations_turn, self.duration)

    def vibrate_left_deviation(self):
        if self.updown:
            self.vibrate_pattern(self.vibrator3_pub, self.num_vibrations_deviation, self.duration)
        else:
            self.vibrate_pattern(self.vibrator4_pub, self.num_vibrations_deviation, self.duration)

    def vibrate_right_deviation(self):
        if self.updown:
            self.vibrate_pattern(self.vibrator4_pub, self.num_vibrations_deviation, self.duration)
        else:
            self.vibrate_pattern(self.vibrator3_pub, self.num_vibrations_deviation, self.duration)

    def vibrate_front(self):
        self.vibrate_pattern(self.vibrator1_pub, self.num_vibrations_confirmation,
                             self.duration_single_vibration)

    def vibrate_about_left_turn(self):
        if self.updown:
            self.vibrate_pattern(
                self.vibrator3_pub, self.num_vibrations_about_turn, self.duration_about_turn)
        else:
            self.vibrate_pattern(
                self.vibrator4_pub, self.num_vibrations_about_turn, self.duration_about_turn)

    def vibrate_about_right_turn(self):
        if self.updown:
            self.vibrate_pattern(
                self.vibrator4_pub, self.num_vibrations_about_turn, self.duration_about_turn)
        else:
            self.vibrate_pattern(
                self.vibrator3_pub, self.num_vibrations_about_turn, self.duration_about_turn)

    def vibrate_back(self):
        self.vibrate_pattern(self.vibrator2_pub, self.num_vibrations_confirmation,
                             self.duration_single_vibration)

    def vibrate_button_click(self):
        self.vibrate_pattern(
            self.vibrator1_pub, self.num_vibrations_button_click, self.duration_button_click)

    def vibrate_button_holddown(self):
        self.vibrate_pattern(
            self.vibrator1_pub, self.num_vibrations_button_holddown, self.duration_button_holddown)

    def vibrate_pattern(self, vibrator_pub, number_vibrations, duration):
        i = 0
        mode = "NEW"

        if mode == "SAFETY":
            while True:
                for v in range(0, duration):
                    self.vibrate(vibrator_pub)
                    time.sleep(0.01)
                self.stop(vibrator_pub)
                self.stop(vibrator_pub)
                self.stop(vibrator_pub)

                i += 1
                if i >= number_vibrations:
                    break
                time.sleep(self.sleep/1000.0)

            # make sure it stops
            for v in range(0, 10):
                self.stop(vibrator_pub)

        if mode == "SIMPLE":
            for i in range(0, number_vibrations):
                self.vibrate(vibrator_pub)
                time.sleep(0.01*duration)
                self.stop(vibrator_pub)
                if i < number_vibrations - 1:
                    time.sleep(self.sleep/1000.0)
            self.stop(vibrator_pub)

        if mode == "NEW":
            for i in range(0, number_vibrations):
                msg = UInt8()
                msg.data = duration
                vibrator_pub.publish(msg)
                time.sleep(0.01*duration)
                if i < number_vibrations - 1:
                    time.sleep(self.sleep/1000.0)

File: cabot/src/cabot/handle_v2.cpp
/*******************************************************************************
 * Copyright (c) 2023  Miraikan and Carnegie Mellon University
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/

#include <vector>
#include <string>
#include <memory>
#include <map>
#include <utility>
#include <thread>
#include <chrono>
#include <algorithm>

#include "handle_v2.hpp"

using namespace std::chrono_literals;

const char* Handle::stimuli_names[] =
{"unknown", "left_turn", "right_turn", "left_dev", "right_dev", "front",
  "left_about_turn", "right_about_turn", "button_click", "button_holddown"};
const rclcpp::Duration Handle::double_click_interval_ = rclcpp::Duration(0, 250000000);
const rclcpp::Duration Handle::ignore_interval_ = rclcpp::Duration(0, 50000000);
const rclcpp::Duration Handle::holddown_interval_ = rclcpp::Duration(3, 0);

std::string Handle::get_name(int stimulus)
{
  return stimuli_names[stimulus];
}

Handle::Handle(
  std::shared_ptr<CaBotHandleV2Node> node,
  std::function<void(const std::map<std::string, std::string> &)> eventListener,
  const std::vector<std::string> & buttonKeys)
: node_(node), eventListener_(std::move(eventListener)), buttonKeys_(buttonKeys),
      logger_(rclcpp::get_logger("handle"))
{
  power_ = 255;
  vibrator1_pub_ = node_->create_publisher<std_msgs::msg::UInt8>("/cabot/vibrator1", 100);
  vibrator2_pub_ = node_->create_publisher<std_msgs::msg::UInt8>("/cabot/vibrator2", 100);
  vibrator3_pub_ = node_->create_publisher<std_msgs::msg::UInt8>("/cabot/vibrator3", 100);
  vibrator4_pub_ = node_->create_publisher<std_msgs::msg::UInt8>("/cabot/vibrator4", 100);
  for (int i = 0; i < 9; ++i) {
    last_up[i] = rclcpp::Time(0, 0, RCL_ROS_TIME);
    last_dwn[i] = rclcpp::Time(0, 0, RCL_ROS_TIME);
    up_count[i] = 0;
    btn_dwn[i] = false;
  }
  for (int i = 1; i <= static_cast<int>(ButtonType::BUTTON_CENTER); ++i) {
    std::function<void(const std_msgs::msg::Bool::SharedPtr)> callback =
      [this, i](const std_msgs::msg::Bool::SharedPtr msg) {
        buttonCallback(msg, static_cast<ButtonType>(i));
      };
    button_subs[i] = node_->create_subscription<std_msgs::msg::Bool>(
      "/cabot/pushed_" + std::to_string(button_keys(i)), rclcpp::SensorDataQoS(),
      [this, i](const std_msgs::msg::Bool::SharedPtr msg) {
        buttonCallback(msg, static_cast<ButtonType>(i));
      });
  }
  event_sub_ = node_->create_subscription<std_msgs::msg::String>(
    "/cabot/event", rclcpp::SensorDataQoS(), [this](const std_msgs::msg::String::SharedPtr msg) {
      eventCallback(msg);
    });
  duration_ = 150;
  duration_single_vibration_ = 400;
  duration_about_turn_ = 400;
  duration_button_click_ = 50;
  duration_button_holddown_ = 100;
  sleep_ = 150;
  num_vibrations_turn_ = 4;
  num_vibrations_deviation_ = 2;
  num_vibrations_about_turn_ = 2;
  num_vibrations_confirmation_ = 1;
  num_vibrations_button_click_ = 1;
  num_vibrations_button_holddown_ = 1;
  callbacks_.resize(std::size(stimuli_names), nullptr);
  callbacks_[1] = [this]() {
      vibrateLeftTurn();
    };
  callbacks_[2] = [this]() {
      vibrateRightTurn();
    };
  callbacks_[3] = [this]() {
      vibrateLeftDeviation();
    };
  callbacks_[4] = [this]() {
      vibrateRightDeviation();
    };
  callbacks_[5] = [this]() {
      vibrateFront();
    };
  callbacks_[6] = [this]() {
      vibrateAboutLeftTurn();
    };
  callbacks_[7] = [this]() {
      vibrateAboutRightTurn();
    };
  callbacks_[8] = [this]() {
      vibrateButtonClick();
    };
  callbacks_[9] = [this]() {
      vibrateButtonHolddown();
    };
  vibration_timer_ = node_->create_wall_timer(0.01s, std::bind(&Handle::timer_callback, this));
}

void Handle::timer_callback()
{
  if (vibration_queue_.size() > 0) {
    Vibration & vibration = vibration_queue_.front();
    RCLCPP_DEBUG(rclcpp::get_logger("handle"), "vibration.i = %d", vibration.i);
    RCLCPP_DEBUG(
      rclcpp::get_logger(
        "handle"), "vibration.numberVibration = %d", vibration.numberVibrations);
    if (vibration.numberVibrations == 0) {
      vibration_queue_.erase(vibration_queue_.begin());
      RCLCPP_INFO(rclcpp::get_logger("handle"), "Done");
    } else if (vibration.i == 0 && vibration.numberVibrations > 0) {
      std_msgs::msg::UInt8 msg;
      msg.data = vibration.duration * 0.1;
      vibration.vibratorPub->publish(msg);
      RCLCPP_INFO(rclcpp::get_logger("handle"), "publish %d", vibration.duration);
      RCLCPP_INFO(rclcpp::get_logger("handle"), "sleep %d ms", vibration.duration);
      vibration.i++;
    } else if (vibration.i == vibration.duration * 0.1  && vibration.numberVibrations == 1) {
      vibration.numberVibrations = 0;
    } else if (vibration.i < (vibration.duration + vibration.sleep) * 0.1) {
      vibration.i++;
    } else {
      vibration.i = 0;
      vibration.numberVibrations--;
    }
  }
}

void Handle::buttonCallback(const std_msgs::msg::Bool::SharedPtr msg, int index)
{
  if (index >= 1 && index <= 5) {
    buttonCheck(msg, index);
  }
}

void Handle::buttonCheck(const std_msgs::msg::Bool::SharedPtr msg, int index)
{
  event.clear();
  rclcpp::Time now = node_->get_clock()->now();
  rclcpp::Time zerotime(0, 0, RCL_ROS_TIME);
  if (msg->data && !btn_dwn[index] &&
    !(last_up[index] != zerotime && now - last_up[index] < ignore_interval_))
  {
    event.insert(std::make_pair("button", std::to_string(button_keys(index))));
    event.insert(std::make_pair("up", "False"));
    btn_dwn[index] = true;
    last_dwn[index] = now;
  }
  if (!msg->data && btn_dwn[index]) {
    event.insert(std::make_pair("button", std::to_string(button_keys(index))));
    event.insert(std::make_pair("up", "True"));
    up_count[index]++;
    last_up[index] = now;
    btn_dwn[index] = false;
  }
  if (last_up[index] != zerotime &&
    !btn_dwn[index] &&
    now - last_up[index] > double_click_interval_)
  {
    if (last_dwn[index] != zerotime) {
      event.insert(std::make_pair("buttons", std::to_string(button_keys(index))));
      event.insert(std::make_pair("count", std::to_string(up_count[index])));
    }
    last_up[index] = zerotime;
    up_count[index] = 0;
  }
  if (msg->data && btn_dwn[index] &&
    last_dwn[index] != zerotime &&
    now - last_dwn[index] > holddown_interval_)
  {
    event.insert(std::make_pair("holddown", std::to_string(button_keys(index))));
    last_dwn[index] = zerotime;
  }
  if (!event.empty()) {
    eventListener_(event);
  }
}

void Handle::eventCallback(const std_msgs::msg::String::SharedPtr msg)
{
  const size_t stimuli_names_size = std::size(Handle::stimuli_names);
  const std::string name = msg->data;
  const char** it = std::find(stimuli_names, stimuli_names + stimuli_names_size, name.c_str());
  if(it != stimuli_names + stimuli_names_size){
    int index = std::distance(stimuli_names, it);
    executeStimulus(index);
  }else{
    RCLCPP_DEBUG(logger_, "Stimulus '%s' not found.", name.c_str());
  }
}

void Handle::executeStimulus(int index)
{
  RCLCPP_INFO(logger_, "execute_stimulus, %d", index);
  const std::size_t size = sizeof(stimuli_names) / sizeof(stimuli_names[0]);
  if (index >= 0 && index < static_cast<int>(size) && callbacks_[index]) {
    callbacks_[index]();
    RCLCPP_INFO(logger_, "executed");
  }
}

void Handle::vibrate(rclcpp::Publisher<std_msgs::msg::UInt8>::SharedPtr pub)
{
  std::unique_ptr<std_msgs::msg::UInt8> msg = std::make_unique<std_msgs::msg::UInt8>();
  msg->data = power_;
  pub->publish(std::move(msg));
}

void Handle::stop(rclcpp::Publisher<std_msgs::msg::UInt8>::SharedPtr pub)
{
  std::unique_ptr<std_msgs::msg::UInt8> msg = std::make_unique<std_msgs::msg::UInt8>();
  msg->data = 0;
  pub->publish(std::move(msg));
}

void Handle::vibrateAll(int time)
{
  std_msgs::msg::UInt8 msg;
  msg.data = power_;
  vibrator1_pub_->publish(msg);
  vibrator2_pub_->publish(msg);
  vibrator3_pub_->publish(msg);
  vibrator4_pub_->publish(msg);
  std::this_thread::sleep_for(std::chrono::milliseconds(time));
  msg.data = 0;
  vibrator1_pub_->publish(msg);
  vibrator2_pub_->publish(msg);
  vibrator3_pub_->publish(msg);
  vibrator4_pub_->publish(msg);
}

void Handle::vibrateLeftTurn()
{
  vibratePattern(vibrator3_pub_, num_vibrations_turn_, duration_);
}

void Handle::vibrateRightTurn()
{
  vibratePattern(vibrator4_pub_, num_vibrations_turn_, duration_);
}

void Handle::vibrateLeftDeviation()
{
  vibratePattern(vibrator3_pub_, num_vibrations_deviation_, duration_);
}

void Handle::vibrateRightDeviation()
{
  vibratePattern(vibrator4_pub_, num_vibrations_deviation_, duration_);
}

void Handle::vibrateFront()
{
  vibratePattern(vibrator1_pub_, num_vibrations_confirmation_, duration_single_vibration_);
}

void Handle::vibrateAboutLeftTurn()
{
  vibratePattern(vibrator3_pub_, num_vibrations_about_turn_, duration_about_turn_);
}

void Handle::vibrateAboutRightTurn()
{
  vibratePattern(vibrator4_pub_, num_vibrations_about_turn_, duration_about_turn_);
}

void Handle::vibrateBack()
{
  vibratePattern(vibrator2_pub_, num_vibrations_confirmation_, duration_single_vibration_);
}

void Handle::vibrateButtonClick()
{
  vibratePattern(vibrator1_pub_, num_vibrations_button_click_, duration_button_click_);
}

void Handle::vibrateButtonHolddown()
{
  vibratePattern(vibrator1_pub_, num_vibrations_button_holddown_, duration_button_holddown_);
}

void Handle::vibratePattern(
  rclcpp::Publisher<std_msgs::msg::UInt8>::SharedPtr vibratorPub,
  int numberVibrations, int duration)
{
  int i = 0;
  RCLCPP_INFO(rclcpp::get_logger("handle"), "Start vibratePattern .");
  Vibration vibration;
  vibration.duration = duration;
  vibration.numberVibrations = numberVibrations;
  vibration.sleep = sleep_;
  vibration.vibratorPub = vibratorPub;
  vibration_queue_.push_back(vibration);
}
